web/blog/zeek-in-the-shell/index.md

```diff
@@ -0,0 +1,179 @@
+JSON](https://suricata.readthedocs.io/en/latest/output/eve/eve-json-output.html)
+by default, which is equivalent to Zeek's streaming JSON output, with the
+difference being that Zeek's `_path` field is called `event_type` in Suricata
+logs.
+
+```bash title=suricatify
+#!/bin/sh
+suricata -r /dev/stdin \
+  --set outputs.1.eve-log.filename=/dev/stdout \
+  --set logging.outputs.0.console.enabled=no
+```
+
+Let's break this down:
+
+- The `--set` option take a `name=value` parameter that overrides the settings
+  in your `suricata.yaml` config file.
+- The key `outputs.1.eve-log.filename` refers to the `outputs` array, takes
+  element at index `1`, treats that as object and goes to the nested field
+  `eve-log.filename`. Setting `/dev/stdout` as filename makes Suricata write to
+  stdout.
+- We must set `logging.outputs.0.console.enabled` to `no` because Suricata
+  doesn't write startup log messages on stderr. Since they are no valid JSON, we
+  would otherwise create an invalid JSON output stream.
+
+## User-defined Operators
+
+Now that we have both Zeek and Suricata at our fingertips, how can we work with
+their output more easily? This is where Tenzir comes into play-easy
+[pipelines](/docs/language/pipelines) for security teams to acquire,
+[reshape](/docs/user-guides/get-started), and route event data.
+
+For our use case we look at the [`shell`](/docs/operators/shell) operator in
+more detail: it lifts any Unix command line into a pipeline. Here are two
+examples that rely on our two shell scripts:
+
+```bash
+# Zeek
+zcat pcap.gz | zeekify | tenzir \
+  'read zeek-json
+   | where #type == "zeek.conn"
+   | summarize n=count_distinct(id.orig_h) by id.resp_h
+   | sort n desc'
+# Suricata
+zcat pcap.gz | suricatify | tenzir \
+  'read suricata
+   | where #type == "suricata.flow"
+   | summarize n=count_distinct(src_ip) by dst_ip
+   | sort n desc'
+```
+
+The two aggregations actually do the same thing: count the number of unique
+source IP addresses per destination IP address.
+
+It's a bit unwiedly to write such a command line that requires an external shell
+script to work. This where [user-defined operators](/operators/user-defined)
+come into play. You can write a custom `zeek` and `suricata` operator and ditch
+the shell script:
+
+```yaml title="tenzir.yaml"
+tenzir:
+  operators:
+    zeek: >
+     shell "zeek -r - LogAscii::output_to_stdout=T \
+            JSONStreaming::disable_default_logs=T \
+            JSONStreaming::enable_log_rotation=F \
+            json-streaming-logs"
+     | parse zeek-json
+    suricata: >
+     shell "suricata -r /dev/stdin \
+            --set outputs.1.eve-log.filename=/dev/stdout \
+            --set logging.outputs.0.console.enabled=no"
+     | parse suricata
+```
+
+The difference stands out when you look now at the pipeline definition:
+
+```text title=Zeek
+zeek
+| where #type == "zeek.conn"
+| summarize n=count_distinct(id.orig_h) by id.resp_h
+| sort n desc
+```
```

Comment:
Does this actually work as intended? `shell` is both `bytes -> bytes` and `void -> bytes`. We do not add an implicit `load -` if the pipeline can be instantiated for `void`. 

You might be able to resolve this by adding `load -` to the alias definition, but then it cannot be used anymore as an inner operator. I think we don't have a clean solution for this at the moment.

Reply:
My comment is not about the output type of `zeek` or `shell`, but rather its input type. The pipeline above is a well-formed `void -> events` pipeline (and also a well-formed `bytes -> events` pipeline, but we "prefer" the `void -> events` type in `exec.cpp`). Unless you prefix it with another operator, we will thus not add an implicit `load -`, and your `shell` implementation is therefore instantiated for `void`. I might be mistaken, but I think this means that we do not forward the stdin of the Tenzir process to `shell`.

Reply:
Ah, so the fix would be to make it a transform by appending `load -` in the UDO definition?

Reply:
That would work for your examples, but then you of course can't use it anymore as an inner (non-source) operator. The only alternatives that I can think of are changing `shell_operator::operator()(operator_control_plane&)` to steal `stdin` (but I'm not sure whether we want to do that), or prefer `bytes -> void` in `exec.cpp` (such that we add an implicit `load -` if the pipeline begins with `shell` or other operators that can act both as a source and as a transformation).

Reply:
I think I get it now, but I'm too far away from the code in the operator control plane. @dominiklohmann thoughts on how to proceed here? Should we factor this into a separate issue?

As user, I would like UDOs to work transparently, i.e., use them as source and transform as in this blog. Let's try to aim for this UX.

Reply:
There are two ways to get that to work:
1. Change the `void -> bytes` version of `shell` to get the entire stdin of the parent process.
2. Remove the `void -> bytes` version of `shell` and always rely on the implicit `load -` source.

This is not an easy call to make:
- Option (1) will make the void -> bytes version of the operator clash with @mavam's `tui` operator, because only one operator may use stdin.
- For deployed pipelines via the pipeline manager API, you don't get implicit sources and sinks, so option (2) would require adding an explicit `load file -` for deployed pipelines that start with `shell` currently.

I think (2) is the way to go. It's a fast change to make, but also a breaking change in the API, so we should make it on the sooner side.

Reply:
But (2) + implicit source has the same effect as (1) in that it breaks the `tui` operator. Yes, I can work around that with `load file /dev/null`, but I can also do that when we keep it. So the only real difference is that I have now have to add `load file -` when using the explorer.

Reply:
At the same time: When using the pipeline manager, I feel like you don't want to grab `stdin` (why would you do that?). So what  preferring `bytes -> void` over `void -> void` in `exec.cpp`, and keeping the `shell` overloads as-is? As written above, I think that should work. Are there any negative side effects from this?

Reply:
> But (2) + implicit source has the same effect as (1) in that it breaks the `tui` operator.

Not sure how I missed that, you are absolutely correct.


> So what preferring `bytes -> void` over `void -> void` in `exec.cpp`,

I don't think this is a good solution. If a pipeline can be considered closed then we should not be touching it.

I think that the current behavior is the behavior we want to have: If the user wants `shell` to read from stdin of the `tenzir` process, then they must explicitly say so with `load file -`.

Reply:
I just noticed that you could also add a `pass` in front of the user-defined operator definition to enforce that this shell operator still needs some input... In this situation, I think this would lead to the perfect behavior, but it's quite hacky.

But yea, it's a difficult situation.

One more option (as if 4 options were not enough): We could also detect whether `stdin` is connected to a pipe to give `foo | tenzir 'shell "bar" | baz` the right behavior.

Reply:
Doing any form of TTY detection is a pain for people using our software through Docker, so we should be very careful about that. I would like to avoid it unless we absolutely must.

Reply:
Well, then I fear we are out of options here. The only solution that checks all boxes is adding the `load file -` explicitly, but you can try out the effect on the pipelines in this post. I don't like it too much.

IMHO the best of all these non-optimal solutions is preferring `bytes -> void` in `exec.cpp`, as this should just "do the right thing", albeit at the cost of a small impurity. It's also how the shell itself works when using pipelines. But I'll let you make the call.

Reply:
> Well, then I fear we are out of options here. The only solution that checks all boxes is adding the `load file -` explicitly, but you can try out the effect on the pipelines in this post. I don't like it too much.

The consequence would be that you cannot use it as transformation anymore, or combine with a different loader. I think that's a big limitation because it breaks composition—which is what operators are all about.

Reply:
Jannis and I just had a call on this and agreed to make the following change: We want to let operators know whether they are allowed to acquire stdin.

We will then allow this for pipelines started via `tenzir <pipeline>`, and forbid it for all other things that start pipelines. The void-to-bytes `shell` operator changes its behavior based on this flag.

Jannis will implement this before the targeted release date of this blog post in a separate PR.

Reply:
This now works as expected, I think we can resolve this thread.

Reply:
Great. Did you test it yourself?

Reply:
(Feel free to resolve if you think it's sufficiently addressed.)

Reply:
We tested that the `shell` operator should now behave as required when in source position, but not the concrete example in this post. If this suffices for you, please resolve.